src/box/iproto.cc

```diff
@@ -2247,76 +2318,112 @@ iproto_do_cfg_f(struct cbus_call_msg *m)
 {
 	struct iproto_cfg_msg *cfg_msg = (struct iproto_cfg_msg *) m;
 	int old;
+	struct iproto_thread *iproto_thread = cfg_msg->iproto_thread;
+
 	try {
 		switch (cfg_msg->op) {
 		case IPROTO_CFG_MSG_MAX:
-			cpipe_set_max_input(&tx_pipe,
+			cpipe_set_max_input(&iproto_thread->tx_pipe,
 					    cfg_msg->iproto_msg_max / 2);
 			old = iproto_msg_max;
 			iproto_msg_max = cfg_msg->iproto_msg_max;
 			if (old < iproto_msg_max)
-				iproto_resume();
+				iproto_resume(iproto_thread);
 			break;
 		case IPROTO_CFG_LISTEN:
-			if (evio_service_is_active(&binary))
-				evio_service_stop(&binary);
-			if (cfg_msg->uri != NULL &&
-			    (evio_service_bind(&binary, cfg_msg->uri) != 0 ||
-			     evio_service_listen(&binary) != 0))
+			if (evio_service_listen(&iproto_thread->binary) != 0)
 				diag_raise();
-			cfg_msg->addrlen = binary.addr_len;
-			cfg_msg->addr = binary.addrstorage;
+			break;
+		case IPROTO_CFG_STOP:
+			if (evio_service_is_active(&iproto_thread->binary))
+				evio_service_stop(&iproto_thread->binary);
 			break;
 		default:
 			unreachable();
 		}
 	} catch (Exception *e) {
 		return -1;
 	}
+
 	return 0;
 }
 
 static inline void
 iproto_do_cfg(struct iproto_cfg_msg *msg)
 {
-	if (cbus_call(&net_pipe, &tx_pipe, msg, iproto_do_cfg_f,
-		      NULL, TIMEOUT_INFINITY) != 0)
-		diag_raise();
+	for( auto& iproto_thread : iproto_threads ) {
+		msg->iproto_thread = iproto_thread;
+		if (cbus_call(&iproto_thread->net_pipe,
+			      &iproto_thread->tx_pipe, msg,
+			      iproto_do_cfg_f,
+			      NULL, TIMEOUT_INFINITY) != 0)
+			diag_raise();
+	}
 }
 
 void
 iproto_listen(const char *uri)
 {
+	struct evio_service binary;
 	struct iproto_cfg_msg cfg_msg;
-	iproto_cfg_msg_create(&cfg_msg, IPROTO_CFG_LISTEN);
-	cfg_msg.uri = uri;
+
+	memset(&binary, 0, sizeof(binary));
+	iproto_cfg_msg_create(&cfg_msg, IPROTO_CFG_STOP);
 	iproto_do_cfg(&cfg_msg);
-	iproto_bound_address_storage = cfg_msg.addr;
-	iproto_bound_address_len = cfg_msg.addrlen;
+	if (uri != NULL) {
+		if (evio_service_bind(&binary, uri) != 0)
+			diag_raise();
+		for( auto& iproto_thread : iproto_threads ) {
```

Comment:
I don't like this solutions. Actually I don't know how to do it good, but I didn't think a lot. I see the following problems to be fixed:
1. You set here `iproto_thread's` members and it's not obvious why it is safe. Please don't explain me, I understand, but you should make the code understandable by itself.
2. This new meesage `IPROTO_CFG_STOP` is not documented fine. It seems that doesn't require the service to be started. At the same time `IPROTO_CFG_LISTEN` requires (!) service to be stopped. That's very confusing.
3. You do a lot work right here, while doing a lot if work in `IPROTO_CFG_LISTEN` handler. That's not obvious either.